File: 1389/create_target_array_in_the_given_order.go
// 1389. Create Target Array in the Given Order
// https://leetcode.com/problems/create-target-array-in-the-given-order/
package main

import (
	"fmt"

	"github.com/k0kubun/pp"
)

func main() {
	pp.Println("=========================================")
	fmt.Println(createTargetArray([]int{0, 1, 2, 3, 4}, []int{0, 1, 2, 2, 1}))
	fmt.Println([]int{0, 4, 1, 3, 2})
	pp.Println("=========================================")
	fmt.Println(createTargetArray([]int{1, 2, 3, 4, 0}, []int{0, 1, 2, 3, 0}))
	fmt.Println([]int{0, 1, 2, 3, 4})
	pp.Println("=========================================")
	fmt.Println(createTargetArray([]int{1}, []int{0}))
	fmt.Println([]int{1})
	pp.Println("=========================================")
	fmt.Println(createTargetArray([]int{}, []int{}))
	fmt.Println([]int{})
	pp.Println("=========================================")
	fmt.Println(createTargetArray([]int{1, 2, 3, 4, 5}, []int{0, 1, 2, 3, 4}))
	fmt.Println([]int{1, 2, 3, 4, 5})
	pp.Println("=========================================")
	fmt.Println(createTargetArray([]int{4, 2, 4, 3, 2}, []int{0, 0, 1, 3, 1}))
	fmt.Println([]int{2, 2, 4, 4, 3})
	pp.Println("=========================================")
}

// 1 <= nums.length, index.length <= 100
// nums.length == index.length
// 0 <= nums[i] <= 100
// 0 <= index[i] <= i
func createTargetArray(nums []int, indices []int) []int {
	targetArray := make([]int, len(nums))
	for i, index := range indices {
		copy(targetArray[index+1:i+1], targetArray[index:i])
		targetArray[index] = nums[i]
	}
	return targetArray
}

File: 1389/create_target_array_in_the_given_order.js
// 1389. Create Target Array in the Given Order
// https://leetcode.com/problems/create-target-array-in-the-given-order/

/**
 * https://leetcode.com/problems/create-target-array-in-the-given-order/discuss/547440/JavaScript-Update-indices-recursively-without-splice
 */

const main = () => {
  console.log("=========================================");
  console.log(createTargetArray([0, 1, 2, 3, 4], [0, 1, 2, 2, 1]));
  console.log([0, 4, 1, 3, 2]);
  console.log("=========================================");
  console.log(createTargetArray([1, 2, 3, 4, 0], [0, 1, 2, 3, 0]));
  console.log([0, 1, 2, 3, 4]);
  console.log("=========================================");
  console.log(createTargetArray([1], [0]));
  console.log([1]);
  console.log("=========================================");
  console.log(createTargetArray([], []));
  console.log([]);
  console.log("=========================================");
  console.log(createTargetArray([1, 2, 3, 4, 5], [0, 1, 2, 3, 4]));
  console.log([1, 2, 3, 4, 5]);
  console.log("=========================================");
  console.log(createTargetArray([4, 2, 4, 3, 2], [0, 0, 1, 3, 1]));
  console.log([2, 2, 4, 4, 3]);
  console.log("=========================================");
};

/**
 * @param {number[]} nums
 * @param {number[]} index
 * @return {number[]}
 */
function createTargetArray(nums, index) {
  const indices = Array(index.length).fill(null);
  for (const [from, to] of index.entries()) {
    // console.debug('=== DEBUG START ======================================');
    // console.debug("from:", from, "to:", to);
    set(from, to);
    // console.debug(indices);
    // console.debug('=== DEBUG END ========================================');
  }
  return indices.map(index => nums[index]);

  function set(from, to) {
    if (from === null) {
      return;
    }
    set(indices[to], to + 1);
    indices[to] = from;
  }
}

main();
